### PR TITLE
Perf: unicode-data-names

### PR DESCRIPTION
**Pull request by wismill**
Follow-up of #107.

Added comparison to ICU.

### CODE REVIEW

**Comment by wismill** [comment]
Results (Linux x64, 8 × AMD Ryzen 5 2500U):

```
All
  Unicode.Char.General.Names
    name
      String: OK (2.61s)
        40.4 ms ± 318 μs, 155 MB allocated, 1.6 KB copied,  35 MB peak memory, 36% less than baseline
    correctedName
      String: OK (0.68s)
        44.7 ms ± 744 μs, 155 MB allocated, 1.7 KB copied,  35 MB peak memory, 34% less than baseline
    nameOrAlias
      String: OK (20.75s)
        40.6 ms ± 504 μs, 155 MB allocated, 1.5 KB copied,  35 MB peak memory, 36% less than baseline
    nameAliasesByType
      String: OK (6.83s)
        109  ms ± 1.4 ms, 518 MB allocated, 3.0 KB copied,  35 MB peak memory, 21% less than baseline
    nameAliasesWithTypes
      String: OK (1.19s)
        18.8 ms ± 302 μs,  52 MB allocated, 343 B  copied,  35 MB peak memory, 58% less than baseline
    nameAliases
      String: OK (1.97s)
        15.4 ms ± 182 μs,  51 MB allocated, 285 B  copied,  35 MB peak memory,  9% less than baseline
```

File: unicode-data-names/unicode-data-names.cabal
Comment (harendra-kumar) [inline unicode-data-names/unicode-data-names.cabal:119]: Non-standard paths should not be hardcoded here. We can use cabal.project to specify these paths.

**Comment by harendra-kumar** [comment]
How does it compare to ICU?

File: unicode-data-names/unicode-data-names.cabal
Comment (wismill) [reply unicode-data-names/unicode-data-names.cabal:119]: I agree. Those are taken from the `text-icu` package, I do not know what use instead. Should I remove them completely and mark it as non buildable for this platform?

**Comment by wismill** [comment]
> How does it compare to ICU?

About 5 times faster. However I did not check much the implementation. It does _not_ rely on a reliable library as `text-icu`: I added the cbits first to enable testing. Benchmark give us a rough idea; else we had nothing to compare to!

**Comment by wismill** [comment]
@harendra-kumar Any idea how to allow failure of GHC-head in the CI? Only results I got are ongoing discussions to implement the feature but no temporary fix. 

How could we mark it as always succeeding? We should then always check the details.

File: unicode-data-names/unicode-data-names.cabal
Comment (harendra-kumar) [inline unicode-data-names/unicode-data-names.cabal:112]: Should also use optimization option, `-O2` should be fine or maybe `-Ofast`.

File: unicode-data-names/unicode-data-names.cabal
Comment (harendra-kumar) [reply unicode-data-names/unicode-data-names.cabal:119]: We can leave it buildable, if the library is found in standard path it might build, if not the onus is on the user to provide the installation and use cabal options to specify where to find it, we can add these as suggested paths in the README.

**Comment by harendra-kumar** [comment]
> How could we mark it as always succeeding? We should then always check the details.

Use `ignore_error: true`, like [here](https://github.com/composewell/streamly/blob/8fe249da3be83b3f3ea301ed541895f2350d6846/.github/workflows/haskell.yml#L76) . Note, you will have to add `ignore_error` to all other CIs as false.

File: unicode-data-names/unicode-data-names.cabal
Comment (wismill) [reply unicode-data-names/unicode-data-names.cabal]: It is not mandatory, it will be added with e.g. `cabal -O2 run`.

File: unicode-data-names/unicode-data-names.cabal
Comment (wismill) [reply unicode-data-names/unicode-data-names.cabal]: Ok. I moved `icu-names` to its own (internal) package. In fact there is no reason to ship it with `unicode-data-names`. It is really only an internal package, enabling further tests.

**Comment by wismill** [comment]
I updated the CI. We should probably use `fail-fast: true`, shouldn’t we?

**Comment by harendra-kumar** [comment]
> I updated the CI. We should probably use `fail-fast: true`, shouldn’t we?

It will save on resources, by cancelling all CIs on first error, but it will not show you errors in all CIs.

**Comment by wismill** [comment]
I added 2 optional APIs: `ByteString` and `Text`. The `ByteString` one is slightly less performant than `Text`, which I did not expect. I guess it is good enough for a first version.

Latests benchmarks:

```
All
  Unicode.Char.General.Names
    name
      String
        unicode-data: OK (5.28s)
          41.1 ms ± 597 μs, 155 MB allocated, 2.7 KB copied,  35 MB peak memory
        icu:          OK (1.43s)
          204  ms ± 7.3 ms, 524 MB allocated, 8.4 KB copied,  51 MB peak memory, 4.96x
      ByteString
        unicode-data: OK (0.82s)
          26.3 ms ± 496 μs,  68 MB allocated, 1.0 KB copied,  51 MB peak memory
      Text
        unicode-data: OK (1.55s)
          24.4 ms ± 685 μs,  64 MB allocated, 875 B  copied,  51 MB peak memory
        icu:          OK (4.02s)
          129  ms ± 2.1 ms, 277 MB allocated, 3.7 KB copied,  51 MB peak memory, 5.30x
    correctedName
      String
        unicode-data: OK (0.38s)
          54.7 ms ± 2.1 ms, 151 MB allocated, 3.1 KB copied,  51 MB peak memory
        icu:          OK (4.76s)
          316  ms ± 6.2 ms, 729 MB allocated, 9.6 KB copied,  52 MB peak memory, 5.78x
      ByteString
        unicode-data: OK (0.95s)
          30.4 ms ± 848 μs,  68 MB allocated, 957 B  copied,  52 MB peak memory
      Text
        unicode-data: OK (3.66s)
          28.8 ms ± 872 μs,  64 MB allocated, 741 B  copied,  52 MB peak memory
        icu:          OK (1.71s)
          245  ms ± 8.9 ms, 475 MB allocated, 6.2 KB copied,  52 MB peak memory, 8.50x
    nameAliasesByType
      String
        unicode-data: OK (1.06s)
          149  ms ± 4.9 ms, 514 MB allocated, 6.3 KB copied,  52 MB peak memory
      ByteString
        unicode-data: OK (2.28s)
          150  ms ± 3.7 ms, 516 MB allocated, 5.7 KB copied,  52 MB peak memory
      Text
        unicode-data: OK (0.45s)
          148  ms ± 3.6 ms, 508 MB allocated, 9.3 KB copied,  52 MB peak memory
    nameAliasesWithTypes
      String
        unicode-data: OK (1.50s)
          23.7 ms ± 713 μs,  52 MB allocated, 577 B  copied,  52 MB peak memory
      ByteString
        unicode-data: OK (0.36s)
          23.3 ms ± 774 μs,  51 MB allocated, 821 B  copied,  52 MB peak memory
      Text
        unicode-data: OK (1.47s)
          23.2 ms ± 329 μs,  52 MB allocated, 558 B  copied,  52 MB peak memory
    nameAliases
      String
        unicode-data: OK (1.21s)
          18.8 ms ± 367 μs,  51 MB allocated, 531 B  copied,  52 MB peak memory
      ByteString
        unicode-data: OK (4.66s)
          18.3 ms ± 497 μs,  51 MB allocated, 452 B  copied,  52 MB peak memory
      Text
        unicode-data: OK (2.41s)
          18.8 ms ± 721 μs,  51 MB allocated, 460 B  copied,  52 MB peak memory
```

**Comment by wismill** [comment]
@harendra-kumar I am statisfied with the current state, although there are some points that surprise me:
- `Text` is faster than `ByteString` for `name`, `nameOrAlias` and `correctedName`. 
- `ByteString` < `Text` < `String`  for `nameAliases` and `nameAliasesWithTypes` (expected), but the 3 perform almost equal for `nameAliasesByType`.

Anyway, the perf is already good and I spent a good amount of time on this, so if there is no suggestion to improve these, I think we can merge as it is.

Latest benchmarks:

```
All
  Unicode.Char.General.Names
    name
      String
        unicode-data: OK (0.60s)
          192  ms ± 3.5 ms, 804 MB allocated,  11 KB copied,  89 MB peak memory, 21% less than baseline
        icu:          OK (2.05s)
          662  ms ±  15 ms, 2.1 GB allocated,  16 KB copied, 112 MB peak memory, 3.45x
      ByteString
        unicode-data: OK (0.81s)
          51.5 ms ± 1.9 ms, 154 MB allocated, 1.1 KB copied, 115 MB peak memory
      Text
        unicode-data: OK (0.32s)
          39.3 ms ± 1.5 ms, 118 MB allocated, 1.4 KB copied, 120 MB peak memory
        icu:          OK (1.36s)
          189  ms ± 2.8 ms, 380 MB allocated, 2.6 KB copied, 120 MB peak memory, 4.82x

All
  Unicode.Char.General.Names
    correctedName
      String
        unicode-data: OK (0.60s)
          193  ms ± 5.4 ms, 804 MB allocated,  63 KB copied,  89 MB peak memory, 22% less than baseline
        icu:          OK (2.34s)
          772  ms ±  11 ms, 2.3 GB allocated,  18 KB copied, 112 MB peak memory, 4.01x
      ByteString
        unicode-data: OK (1.81s)
          55.8 ms ± 2.0 ms, 157 MB allocated, 986 B  copied, 112 MB peak memory
      Text
        unicode-data: OK (1.45s)
          44.1 ms ± 891 μs, 125 MB allocated, 810 B  copied, 120 MB peak memory
        icu:          OK (4.53s)
          295  ms ± 6.5 ms, 571 MB allocated, 3.2 KB copied, 120 MB peak memory, 6.69x

All
  Unicode.Char.General.Names
    nameOrAlias
      String
        unicode-data: OK (0.64s)
          197  ms ± 4.9 ms, 804 MB allocated,  10 KB copied,  89 MB peak memory, 22% less than baseline
      ByteString
        unicode-data: OK (0.47s)
          57.3 ms ± 1.4 ms, 150 MB allocated, 1.4 KB copied, 112 MB peak memory
      Text
        unicode-data: OK (0.77s)
          48.2 ms ± 1.4 ms, 122 MB allocated, 875 B  copied, 119 MB peak memory

All
  Unicode.Char.General.Names
    nameAliasesWithTypes
      String
        unicode-data: OK (2.28s)
          147  ms ± 2.0 ms, 643 MB allocated, 7.6 KB copied,  87 MB peak memory, 16% less than baseline
      ByteString
        unicode-data: OK (2.16s)
          66.6 ms ± 2.3 ms, 248 MB allocated, 1.4 KB copied, 101 MB peak memory
      Text
        unicode-data: OK (1.52s)
          96.7 ms ± 2.1 ms, 286 MB allocated, 1.5 KB copied, 101 MB peak memory

All
  Unicode.Char.General.Names
    nameAliasesByType
      String
        unicode-data: OK (1.00s)
          134  ms ± 2.8 ms, 541 MB allocated, 6.2 KB copied,  87 MB peak memory, 47% less than baseline
      ByteString
        unicode-data: OK (2.03s)
          131  ms ± 1.3 ms, 541 MB allocated, 5.7 KB copied, 111 MB peak memory
      Text
        unicode-data: OK (0.42s)
          129  ms ± 3.9 ms, 541 MB allocated, 7.2 KB copied, 111 MB peak memory

All
  Unicode.Char.General.Names
    nameAliases
      String
        unicode-data: OK (1.89s)
          121  ms ± 3.1 ms, 541 MB allocated, 5.8 KB copied,  87 MB peak memory, 11% less than baseline
      ByteString
        unicode-data: OK (0.63s)
          37.5 ms ± 1.4 ms, 146 MB allocated, 1.0 KB copied, 107 MB peak memory
      Text
        unicode-data: OK (1.04s)
          66.4 ms ± 2.1 ms, 186 MB allocated, 1.1 KB copied, 107 MB peak memory
```

**Comment by wismill** [comment]
Well, I gave a final try: going low lovel with primops proved to be more efficient than FFI.

The downside is that if primops or `ByteString` internals change, we will have to adapt. But:
- For `ByteString` internals we rely on package bounds.
- The CI passes, so these primops seem very stable.

These low-level stuff could be applied to `Text` API as well, but let’s call it a day.

New benchmark results, which show `ByteString` and `Text` API  now much closer:

```
All
  Unicode.Char.General.Names
    name
      String
        unicode-data: OK (1.31s)
          182  ms ± 3.2 ms, 813 MB allocated, 8.5 KB copied,  91 MB peak memory, 27% less than baseline
      ByteString
        unicode-data: OK (0.69s)
          40.7 ms ± 1.0 ms, 127 MB allocated, 922 B  copied, 119 MB peak memory
      Text
        unicode-data: OK (0.64s)
          37.5 ms ± 915 μs, 122 MB allocated, 877 B  copied, 119 MB peak memory

All
  Unicode.Char.General.Names
    correctedName
      String
        unicode-data: OK (2.85s)
          186  ms ± 697 μs, 815 MB allocated, 8.3 KB copied,  91 MB peak memory, 28% less than baseline
      ByteString
        unicode-data: OK (0.72s)
          43.5 ms ± 1.4 ms, 127 MB allocated, 920 B  copied, 115 MB peak memory
      Text
        unicode-data: OK (0.70s)
          41.8 ms ± 938 μs, 122 MB allocated, 877 B  copied, 115 MB peak memory

All
  Unicode.Char.General.Names
    nameOrAlias
      String
        unicode-data: OK (1.35s)
          187  ms ± 1.4 ms, 813 MB allocated, 8.7 KB copied,  91 MB peak memory, 24% less than baseline
      ByteString
        unicode-data: OK (0.75s)
          47.0 ms ± 944 μs, 127 MB allocated, 925 B  copied, 117 MB peak memory
      Text
        unicode-data: OK (0.74s)
          44.0 ms ± 1.4 ms, 122 MB allocated, 877 B  copied, 117 MB peak memory
```

File: experimental/icu/icu.cabal
Comment (harendra-kumar) [inline experimental/icu/icu.cabal:67]: Should use some optimization option e.g. `-O2` or `-Ofast`.

File: experimental/icu/icu.cabal
Comment (harendra-kumar) [inline experimental/icu/icu.cabal:46]: Did you try using `-O2` if that makes any impact on perf?

File: unicode-data-names/lib/Unicode/Char/General/Names.hs
Comment (harendra-kumar) [inline unicode-data-names/lib/Unicode/Char/General/Names.hs:44]: My only concern is that we are making the implementation significantly more complex to make these performance gains. Is the performance gain required for unicode-data-names? Are these APIs to be used at places where the additional performance matters? Maybe that is the reason why ICU performance is not so good, because they do not care about the perf of these APIs.

Comment (harendra-kumar) [reply unicode-data-names/lib/Unicode/Char/General/Names.hs:44]: I have the same concern about Text/ByteString APIs. If the performance does not matter just the String APIs are enough. More APIs add significantly more maintenance overhead.

I know I am a bit late about this, you have done a lot of hard work on this. But in general we should optimize only if it matters. I do not have much idea where these APIs are used, so I might be wrong.

File: experimental/icu/icu.cabal
Comment (wismill) [reply experimental/icu/icu.cabal:46]: I always run `cabal` with `-O2`.

File: unicode-data-names/lib/Unicode/Char/General/Names.hs
Comment (wismill) [reply unicode-data-names/lib/Unicode/Char/General/Names.hs:44]: 1. The hard work is already done! It was a good exercise to start with a package with relatively small API. Next come `unicode-data`.
2. _You should not worry about the special cases for the names._ These are specific (huge) ranges of characters related to CJK. Any addition with the same template will be processed automatically, because I designed the generator and the optimisation to require no maintenance effort. Any new template will be treated as standard name; a manual change would be required only if optimisation is worth it; else it will work normally.
3. The implementation is quite similar for the 3 APIs.
4. It is not only about speed perf, but also about library/executable _size_.
5. You suggested from the beginning to use `CString`s. The current code is not a simple lookup any more and requires byte arrays manipulation. So it seems quite natural to use the standard `bytestring` and `text`, doesn’t it?
6. The only maintenance I see is updating at each new Unicode version. The Text and ByteString API rely on internal modules of third-party libs, but these are already considered almost public and seem quite stable.

Comment (wismill) [reply unicode-data-names/lib/Unicode/Char/General/Names.hs:44]: We can also mark the new API as experimental and without guarantee to be maintained.

Comment (wismill) [reply unicode-data-names/lib/Unicode/Char/General/Names.hs:44]: As the author of this package, I would like to have the best performance, but obviously I will not be optimising it forever. This is only the first round after its creation.
I am also the current maintainer, so it is designed with low maintenance effort in mind.

Comment (harendra-kumar) [reply unicode-data-names/lib/Unicode/Char/General/Names.hs:44]: Ok. I went through the code only at a high level. Did not dig deep into details. I am assuming your tests will cover any issues.

Would have been nice if we had a single interface which could be adapted to string/bytestring/text. And the users of this library themselves could do it. We are using internals of bytestring/text so whenever they change we will have to adapt to the change. And it is unlikely that users will use these flags. You are enthusiastic and have time at this point, but at some point of time you will get busier and this will become hard for you to maintain. But I will leave this to your judgement, and I know it can be removed if we cannot maintain it.

**Review by harendra-kumar — APPROVED** [review]

**Comment by wismill** [comment]
Rebased

**Comment by wismill** [comment]
We should use [`text-builder-linear`](https://hackage.haskell.org/package/text-builder-linear) for `Text` and `ByteString`. I am going to leave this at the moment, because this repo really needs some love. Will open an issue to do not forget about it.

**Pull request merged**
